src/ocean/util/container/map/model/BucketElementFreeList.d

```diff
@@ -436,5 +266,186 @@ abstract class IBucketElementFreeList: IAllocator
+
+/*******************************************************************************
+
+    Minimalistic Linked List implementation.
+    Doesn't have internal nodes for store items.
+    Doesn't own stored items.
+    ListNode should have properly next of type ListNode*
+
+*******************************************************************************/
+
+class SimpleLinkedList(ListNode)
+{
+
+    static assert(is(ListNode == struct),
+                  "ListNode type needs to be a struct, which " ~
+                  ListNode.stringof ~ " is not");
+
+    static if(is(typeof(ListNode.next) Next))
+    {
+        static assert(is(Next == ListNode*),
+                      ListNode.stringof ~ ".next needs to be of type " ~
+                      (ListNode*).stringof ~ ", not " ~ Next.stringof);
+    }
+    else
+    {
+        static assert(false, "need " ~ (ListNode*).stringof ~ " " ~
+                             ListNode.stringof ~ ".next");
+    }
+
+
+    ///Pointer on first item of linked list
+    private ListNode* head = null;
+
+    ///Returns true if list is empty otherwise false
+    public bool empty()
+    {
+        return this.head is null;
     }
+
+    /**************************************************************************
+
+        Removes one item from the list and return removed item
+
+        Returns:
+            Removed ListNode, could be null if list is empty
+
+     **************************************************************************/
+
+    public ListNode* pop()
+    {
+        if (this.empty())
+        {
+            return null;
+        }
+
+        auto top = this.head;
+        this.head = this.head.next;
+        return top;
+    }
+
+    /**************************************************************************
+
+        Inserts one ListNode into LinkedList
+
+        Params:
+            node = ListNode for insertion.
+                   Node should be not null.
+                   List shouldn't contain Node.
+
+     **************************************************************************/
+
+    public void push(ListNode* node)
+    in
+    {
+        auto result = this.find((ListNode* x) { return x == node; });
+        verify(result is null, "List should contain a node only once");
+    }
+    body
+    {
+        verify(node !is null, "Cannot use null as valid ListNode");
+        node.next = this.head;
+        this.head = node;
+    }
+
+    /**************************************************************************
+
+        Iterates over LinkedList call predicate foreach ListNode in LinkedList.
+
+
+        Params:
+            predicate = delegate which define node searching criteria
+
+        Returns:
+            return first node match to predicate
+            either
+            return null if there is no node which match to predicate
+
+     **************************************************************************/
+    public ListNode* find( bool delegate(ListNode*) predicate)
```

Comment:
why do you need this.
Linked list search is freaking (if not pointing into a contiguous memory-block).
it should not expose a find primitive.